### PR TITLE
Implement basic lint detecting manual slice copy

### PR DESCRIPTION
**Pull request by marcusklaas**
This lint detects manual sequential memory copies between slices.

Should address https://github.com/rust-lang-nursery/rust-clippy/issues/1831.

### CODE REVIEW

File: tests/ui/for_loop.stderr [hunk 1]
```diff
@@ -506,5 +506,73 @@ help: use the corresponding method
...
+405 | |     }
+    | |_____^ help: try replacing the loop by: `dst.clone_from_slice(&src[10..])`
+
+error: it looks like you're manually copying between slices
+   --> $DIR/for_loop.rs:408:5
+    |
+408 | /     for i in 11..src.len() {
+409 | |         dst[i] = src[i - 10];
+410 | |     }
+    | |_____^ help: try replacing the loop by: `dst[11..].clone_from_slice(&src[(11 + -10)..(src.len() + -10)])`
```
Comment (oli-obk) [inline tests/ui/for_loop.stderr]: I think we should not produce output like `a + -b`, can you try to detect this situation, as it seems to be rather common.

File: tests/ui/for_loop.stderr [hunk 2]
```diff
@@ -506,5 +506,73 @@ help: use the corresponding method
...
+    | |_____^ help: try replacing the loop by: `dst.clone_from_slice(&src[..dst.len()])`
+
+error: it looks like you're manually copying between slices
+   --> $DIR/for_loop.rs:426:5
+    |
+426 | /     for i in 10..256 {
+427 | |         dst[i] = src[i - 5];
+428 | |         dst2[i + 500] = src[i]
+429 | |     }
+    | |_____^ help: try replacing the loop by: `dst[10..].clone_from_slice(&src[(10 + -5)..(256 + -5)])`
```
Comment (oli-obk) [inline tests/ui/for_loop.stderr]: same as above.

File: clippy_lints/src/loops.rs [hunk 3]
```diff
@@ -83,7 +103,8 @@ declare_lint! {
 /// implements `IntoIterator`, so that possibly one value will be iterated,
 /// leading to some hard to find bugs. No one will want to write such code
 /// [except to win an Underhanded Rust
-/// Contest](https://www.reddit.com/r/rust/comments/3hb0wm/underhanded_rust_contest/cu5yuhr).
+/// Contest](https://www.reddit.
+/// com/r/rust/comments/3hb0wm/underhanded_rust_contest/cu5yuhr).
```
Comment (marcusklaas) [inline clippy_lints/src/loops.rs]: Are you sure this affected the `clippy_lints` directory as well? In my experience, `cargo fmt` does not touch these by default..

**Comment by marcusklaas** [comment]
Updated the PR!

File: clippy_lints/src/loops.rs (hunk 3 above)
Comment (oli-obk) [inline clippy_lints/src/loops.rs]: We have a script which runs recursively

File: clippy_lints/src/loops.rs [hunk 4]
```diff
@@ -541,13 +584,266 @@ fn check_for_loop<'a, 'tcx>(
...
+struct FixedOffsetVar {
+    var_name: String,
+    offset: Offset,
+}
+
+fn is_slice_like<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, ty: Ty) -> bool {
+    let is_slice = match ty.sty {
+        ty::TyRef(_, ref subty) => is_slice_like(cx, subty.ty),
+        ty::TySlice(..) => true,
+        ty::TyArray(..) => true,
```
Comment (oli-obk) [inline clippy_lints/src/loops.rs]: rewrite as

```rust
        ty::TySlice(..) | ty::TyArray(..) => true,
```

(this was caught by travis. You can check such things yourself by running `cargo test dogfood`)

File: clippy_lints/src/loops.rs [hunk 5]
```diff
@@ -804,24 +805,41 @@ fn detect_manual_memcpy<'a, 'tcx>(
...
+
+                    // span_lint_and_sugg(
+                    //     cx,
+                    //     MANUAL_MEMCPY,
+                    //     expr.span,
+                    //     "it looks like you're manually copying between slices",
+                    //     "try replacing the loop by",
+                    // format!("{}.clone_from_slice(&{}[{}..{}])", dst, src_var.var_name,
+                    // src_offset, src_limit),
+                    // );
```
Comment (oli-obk) [inline clippy_lints/src/loops.rs]: Please remove the commented out code.

File: clippy_lints/src/loops.rs [hunk 6]
```diff
@@ -804,24 +805,41 @@ fn detect_manual_memcpy<'a, 'tcx>(
...
+                })
+                .join("\n    ");
+
+            if big_sugg.len() > 0 {
```
Comment (oli-obk) [inline clippy_lints/src/loops.rs]: Can you create a test that shows this with multiple suggestions?

Comment (oli-obk) [inline clippy_lints/src/loops.rs]: also, replace it with `!big_sugg.is_empty()`

Comment (marcusklaas) [inline clippy_lints/src/loops.rs]: There is such a test already :-)

Comment (oli-obk) [inline clippy_lints/src/loops.rs]: oh... why didn't a test change after your changes then?

File: tests/ui/for_loop.stderr [hunk 7]
```diff
@@ -506,5 +506,78 @@ help: use the corresponding method
...
+    |
+426 | /     for i in 10..256 {
+427 | |         dst[i] = src[i - 5];
+428 | |         dst2[i + 500] = src[i]
+429 | |     }
+    | |_____^
+    |
+help: try replacing the loop by
+    |
+426 |     dst[10..].clone_from_slice(&src[(10 - 5)..(256 - 5)])
```
Comment (marcusklaas) [inline tests/ui/for_loop.stderr]: This is the test I was referring to.

File: clippy_lints/src/loops.rs (hunk 6 above)
Comment (marcusklaas) [inline clippy_lints/src/loops.rs]: It was in there from the start. The test output did change, though.

**Comment by marcusklaas** [comment]
Thanks for the review @oli-obk!

After a more detailed inspection of the documentation on `clone_from_slice`, I think the current lint suggestions may cause panics. The destination slice must apparently be exactly as long as the source. Also, it may be preferable to use `copy_from_slice` when `T: Copy`, although in my experience `clone_from_slice` also compiles to a memcpy in release builds.

I'll update this PR shortly.

**Comment by oli-obk** [comment]
> Are you sure this affected the clippy_lints directory as well? In my experience, cargo fmt does not touch these by default..

I screwed up even more (I forgot to push the changes!). Please rebase over master now. If you want I can do the rebase, just ping me. Sorry about the trouble.

**Comment by marcusklaas** [comment]
Hey @oli-obk - I'm still getting many style conflicts :-(

What command are you using to do the reformat?

**Comment by oli-obk** [comment]
@marcusklaas you seem to have included my commits in your PR now. I'm not sure how that happened. Try rebasing over the master branch.

> What command are you using to do the reformat?

`cargo fmt`, run in `clippy_lints` and in the root directory

**Comment by llogiq** [comment]
Note that there may be differences between "stable" rustfmt and rustfmt-nightly.

**Comment by oli-obk** [comment]
Our current code base is formatted with stable rustfmt. We could move to rustfmt-nighty. It's not like anyone building clippy doesn't have a nightly lying around somewhere

**Comment by marcusklaas** [comment]
This PR is a mess. https://github.com/rust-lang-nursery/rust-clippy/pull/2021 is a clean new PR.

**Pull request closed without merging**